src/liballoc/allocator.rs

```diff
@@ -0,0 +1,990 @@
+    /// `self`, with a suitable amount of padding between each to
+    /// ensure that each instance is given its requested size and
+    /// alignment. On success, returns `(k, offs)` where `k` is the
+    /// layout of the array and `offs` is the distance between the start
+    /// of each element in the array.
+    ///
+    /// On arithmetic overflow, returns `None`.
+    pub fn repeat(&self, n: usize) -> Option<(Self, usize)> {
+        let padded_size = match self.size.checked_add(self.padding_needed_for(self.align)) {
+            None => return None,
+            Some(padded_size) => padded_size,
+        };
+        let alloc_size = match padded_size.checked_mul(n) {
+            None => return None,
+            Some(alloc_size) => alloc_size,
+        };
+        Some((Layout::from_size_align(alloc_size, self.align), padded_size))
+    }
+
+    /// Creates a layout describing the record for `self` followed by
+    /// `next`, including any necessary padding to ensure that `next`
+    /// will be properly aligned. Note that the result layout will
+    /// satisfy the alignment properties of both `self` and `next`.
+    ///
+    /// Returns `Some((k, offset))`, where `k` is layout of the concatenated
+    /// record and `offset` is the relative location, in bytes, of the
+    /// start of the `next` embedded witnin the concatenated record
+    /// (assuming that the record itself starts at offset 0).
+    ///
+    /// On arithmetic overflow, returns `None`.
+    pub fn extend(&self, next: Self) -> Option<(Self, usize)> {
+        let new_align = cmp::max(self.align, next.align);
+        let realigned = Layout { align: new_align, size: self.size };
+        let pad = realigned.padding_needed_for(next.align);
+        let offset = match self.size.checked_add(pad) {
+            None => return None,
+            Some(offset) => offset,
+        };
+        let new_size = match offset.checked_add(next.size) {
+            None => return None,
+            Some(new_size) => new_size,
+        };
+        Some((Layout::from_size_align(new_size, new_align), offset))
+    }
+
+    /// Creates a layout describing the record for `n` instances of
+    /// `self`, with no padding between each instance.
+    ///
+    /// Note that, unlike `repeat`, `repeat_packed` does not guarantee
+    /// that the repeated instances of `self` will be properly
+    /// aligned, even if a given instance of `self` is properly
+    /// aligned. In other words, if the layout returned by
+    /// `repeat_packed` is used to allocate an array, it is not
+    /// guaranteed that all elements in the array will be properly
+    /// aligned.
+    ///
+    /// On arithmetic overflow, returns `None`.
+    pub fn repeat_packed(&self, n: usize) -> Option<Self> {
+        let size = match self.size().checked_mul(n) {
+            None => return None,
+            Some(scaled) => scaled,
+        };
+        Some(Layout { size: size, align: self.align })
```

Comment:
... I'll make this change, though one might argue that if we can infer at the call-site that the alignment constraint is met, then might it not be better to sidestep the `assert!(align.is_power_of_two())`, rather than hoping that LLVM will optimize it away?